src/arch/wasm.rs

```diff
@@ -0,0 +1,270 @@
+    let d0123lh = i32x4_shuffle::<2, 6, 3, 7>(d01l, d23l);
+    let d4567ll = i32x4_shuffle::<0, 4, 1, 5>(d45l, d67l);
+    let d4567lh = i32x4_shuffle::<2, 6, 3, 7>(d45l, d67l);
+    let d0123hl = i32x4_shuffle::<0, 4, 1, 5>(d01h, d23h);
+    let d0123hh = i32x4_shuffle::<2, 6, 3, 7>(d01h, d23h);
+    let d4567hl = i32x4_shuffle::<0, 4, 1, 5>(d45h, d67h);
+    let d4567hh = i32x4_shuffle::<2, 6, 3, 7>(d45h, d67h);
+
+    // Operating on 64-bits will interleave *consecutive quadruples* of 16-bit integers.
+    data[0] = i64x2_shuffle::<0, 2>(d0123ll, d4567ll);
+    data[1] = i64x2_shuffle::<1, 3>(d0123ll, d4567ll);
+    data[2] = i64x2_shuffle::<0, 2>(d0123lh, d4567lh);
+    data[3] = i64x2_shuffle::<1, 3>(d0123lh, d4567lh);
+    data[4] = i64x2_shuffle::<0, 2>(d0123hl, d4567hl);
+    data[5] = i64x2_shuffle::<1, 3>(d0123hl, d4567hl);
+    data[6] = i64x2_shuffle::<0, 2>(d0123hh, d4567hh);
+    data[7] = i64x2_shuffle::<1, 3>(d0123hh, d4567hh);
+}
+
+#[cfg(target_arch = "wasm32")]
+#[target_feature(enable = "simd128")]
+pub fn dequantize_and_idct_block_8x8(
+    coefficients: &[i16; 64],
+    quantization_table: &[u16; 64],
+    output_linestride: usize,
+    output: &mut [u8],
+) {
+    // The loop below will write to positions [output_linestride * i, output_linestride * i + 8)
+    // for 0<=i<8. Thus, the last accessed position is at an offset of output_linestrade * 7 + 7,
+    // and if that position is in-bounds, so are all other accesses.
+    assert!(
+        output.len()
+            > output_linestride
+                .checked_mul(7)
+                .unwrap()
+                .checked_add(7)
+                .unwrap()
+    );
+
+    const SHIFT: u32 = 3;
+
+    // Read the DCT coefficients, scale them up and dequantize them.
+    let mut data = [i16x8_splat(0); 8];
+    unsafe {
+        for i in 0..8 {
+            data[i] = i16x8_shl(
+                i16x8_mul(
+                    v128_load(coefficients.as_ptr().wrapping_add(i * 8) as *const _),
+                    v128_load(quantization_table.as_ptr().wrapping_add(i * 8) as *const _),
+                ),
+                SHIFT,
+            );
+        }
+    }
+
+    // Usual column IDCT - transpose - column IDCT - transpose approach.
+    idct8(&mut data);
+    transpose8(&mut data);
+    idct8(&mut data);
+    transpose8(&mut data);
+
+    for i in 0..8 {
+        // The two passes of the IDCT algorithm give us a factor of 8, so the shift here is
+        // increased by 3.
+        // As values will be stored in a u8, they need to be 128-centered and not 0-centered.
+        // We add 128 with the appropriate shift for that purpose.
+        const OFFSET: i16 = 128 << (SHIFT + 3);
+        // We want rounding right shift, so we should add (1/2) << (SHIFT+3) before shifting.
+        const ROUNDING_BIAS: i16 = (1 << (SHIFT + 3)) >> 1;
+
+        let data_with_offset = i16x8_add_sat(data[i], i16x8_splat(OFFSET + ROUNDING_BIAS));
+
+        unsafe {
```

Comment:
Could you add a `// SAFETY:` comment explaining why this block is OK? Specifically that the assertion above makes sure the wrapping_add doesn't go out of bounds